### ISSUE TITLE
Library should not contain top level data structures, defeats Closure DCE, use functions instead

### ISSUE THREAD

**Issue by swannodette** [andrewmcveigh/cljs-time#21]
Namespaces like `cljs-time.format` should not use top level data structures. This will immediately defeat Closure dead code elimination. JavaScript objects have the benefit of being optimizable by Google Closure so this pattern is acceptable in Closure itself - this is not true for PersistentHashMaps etc. These top level data structures should be replaced by proper functions.

**Comment by andrewmcveigh** [comment]
Thanks for the advice, I'll be working on this for the next release.

**Comment by Jell** [comment]
any update on that?

**Comment by andrewmcveigh** [comment]
Hey,

I've been testing this in a branch[1], but it's a fairly extensive re-implementation, and I'm not seeing a huge difference in compiled size at the moment.

It's also potentially breaking support for some patterns, so it's not ready for release yet.

[1] https://github.com/andrewmcveigh/cljs-time/blob/exp/closure-dce/src/cljs_time/format.cljs

**Comment by Jell** [comment]
I'm asking because I've been tracking down a bug I found on iPhone 4 that I've been pulling my hair on for 2 days straight, and I don't know the exact details of why but this:

`(time-format/date-parsers "yyyy")`

Returns nil if and only if I run it on an iphone 4 without using the debugger. If I use the debugger it works again.

So I have no idea why, but if I wrap `date-parsers` in a `(defn)` the problem goes away. I suspect it has something to do with the naked top level `let`, as mentioned in this issue. Do you mind if I provide a PR where I wrap all the top level hashes in a `defn`?

**Comment by Jell** [comment]
ah! I just noticed that it's not that easy to do for `date-formatters` for instance... So maybe forget my PR idea :)

**Comment by andrewmcveigh** [comment]
Yeah, the `master` implementation relies on being able to enumerate over `date-parsers`.

Is the issue solely with the iphone4?
Are you calling `(time-format/date-parsers "yyyy")` directly, or is that just the source of the bug?
How are you developing (what environment, editor/IDE, etc)? And which debugger do you mean?
What clojurescript compilation settings are you using?

**Comment by Jell** [comment]
Ah sorry I should be a bit more precise. The issue only happens on iphone 4 with iOS 7, it works on iphone 4S. I am not calling `date-parsers` directly, that's just the source of the bug.

By debugger I mean the safari debugger that allows you to inspect a webpage loaded in the iphone. It's not even a debugger per se, just a web inspector (the only web inspector you have access to when using an iphone).

I'm using the following compilation settings:

```
{:optimizations :whitespace
 :pretty-print true
 :static-fns true}
```

But got the same results using advanced compilation. After looking into it, I could narrow down the error to a much smaller failing case:

```
(def date-parsers
  (let [d :d
        M :M
        y :y
        a :a
        m :m
        s :s
        S :S
        h :h
        H :h]
    {"d" d
     "dd" d
     "D" d
     "DD" d
     "DDD" d
     "dth" d
     "M" M
     "MM" M
     "y" y
     "yy" y
     "yyyy" y
     "Y" y
     "YY" y
     "YYYY" y
     "a" a
     "A" a
     "m" m
     "s" s
     "S" S
     "h" h
     "H" H
     "hh" h}))

(set! (.-onload js/window) (js/alert (date-parsers "yyyy")))
```

This shows an alert with `nil` when running without the webinspector, and `:y` when using the webinspector. My conclusion is that it's not a bug in your lib, but something lower level (probably even a bug in the JS interpreter of the iPhone), but I can circumvent it by not having that top level data structure.

I suspect it has something to do with having some top-level binding that's somehow breaking things. I'll investigate some more and report here and to the ClojureScript devs.

**Comment by Jell** [comment]
I'll also try to make a minimal github project with a setup to reproduce the bug if I can.

**Comment by Jell** [comment]
OK this appears to not be as easily reproducible as I thought, because having only the above code in a freshly created project works. It must be in combination with some other libraries. I will look further into this later this week and try to setup a minimal failing test case, until then you can forget about it! :)

**Comment by andrewmcveigh** [comment]
Thanks for this, if you could produce a failing case that would be great. I've created a new issue [1] for this, as it doesn't seem related to this issue.

[1] https://github.com/andrewmcveigh/cljs-time/issues/40
